Convert convention(tensorflow) functions to pass arguments indirectly.

pschuh opened the pull request:
The purpose of doing this is to make it so that there isn't a difference between levels of abstraction when these are used in a generic context or not. This should be a resolution to SR-8684.

slavapestov commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -883,7 +884,7 @@ class DestructureInputs {
       convention = Convs.getDirect(ownership, forSelf, origParamIndex, origType,
                                    substTL);
-      assert(!isIndirectFormalParameter(convention));
+      // assert(!isIndirectFormalParameter(convention));
```
Can you avoid having to comment out the assert?

slavapestov commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -482,7 +488,16 @@ uint16_t SILGenFunction::emitProlog(ParameterList *paramList,
   resultType = resultType->getCanonicalType(genericSig);
 
-  emitIndirectResultParameters(*this, resultType, DC);
+  if (!emitIndirectResultParameters(*this, resultType, DC)) {
```
Cleaning this up would be nice in general since we want SILGenProlog and SILGenEpilog to handle emitting functions at different levels of abstraction directly. This will let us avoid wrapping literal closures in dynamically allocated thunks when they're passed generically, and such.

slavapestov reviewed (commented):
I don't know what restrictions there are on convention(tensorflow) functions, but in general this is not sufficient. Metatypes and functions are passed differently at different abstraction levels too.

Can a convention(tensorflow) function take a thin or thick function as an argument?

slavapestov commented:
Oh, and tuples too. If a function takes a single tuple argument (eg, `(Int, Int)`), the normal form takes two direct Int arguments; the maximally abstract takes a single indirect `(Int, Int)` argument (not two indirect Int arguments).

It would be great if you could come up with a new way of doing this but in terms of AbstractionPattern to share logic with the rest of reabstraction; then we can use this elsewhere in SILGen to emit closures at the right abstraction level too like I said in my review comment above.

CC @jckarter

pschuh commented:
Hi Slava,

Thanks for taking a look. I was sending this out in order to get early feedback from bgogul (hence the commented out assert). The basic thing here is that we want to be able to emit functions with any abstraction level, and then pin all @convention(tensorflow) functions be maximally abstract so that there will never be any abstraction thunks for @convention(tensorflow).

The first step seems to be making lib/SILGen/SILGenProlog.cpp be more general. You're proposing the next level of generality than what is necessary for solving this bug, so some pointers would be nice.

pschuh commented:
tensorflow functions are very restricted; they can only take things that are convertable to / from tensors as arguments.

jckarter commented:
We already have some convention-specific abstraction handling code because of `@convention(c)` and `@convention(block)`. C function pointers and ObjC blocks can't be reabstracted either, so type lowering knows not to try, and always keeps arguments at the C-representable level of abstraction. This sounds somewhat similar, though with slightly different rules.

slavapestov commented:
Yeah, @jckarter's idea of having `AbstractionPattern::getFunctionParam/ResultType()` seems like it will handle this better.

pschuh committed:
Convert convention(tensorflow) functions to pass arguments indirectly.

The purpose of doing this is to make it so that there isn't a difference
between levels of abstraction when these are used in a generic context
or not.

pschuh commented:
@slavapestov, @jckarter, I was able to switch over to Abstraction Pattern for lib/SIL/SILFunctionType.cpp. I'm not sure I follow what you wanted for lib/SILGen/SILGenProlog.cpp.

bgogul commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -482,7 +488,16 @@ uint16_t SILGenFunction::emitProlog(ParameterList *paramList,
-  emitIndirectResultParameters(*this, resultType, DC);
+  if (!emitIndirectResultParameters(*this, resultType, DC)) {
+    // If we did emit an indirect result above, do so now. This means that we
+    // forced the result to be indirect via the convention.
```
did -> did not?

Also, mark this with SWIFT_ENABLE_TENSORFLOW?

bgogul commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -97,12 +97,17 @@ class EmitBBArguments : public CanTypeVisitor<EmitBBArguments,
     auto parameterInfo = parameters.front();
     parameters = parameters.slice(1);
+
+    if (parameterInfo.isFormalIndirect()) {
```
Mark this as SWIFT_ENABLE_TENSORFLOW?

Can you also add a comment explaining why you need to do this?

bgogul reviewed (approved):
Let us merge this in for now after fixing the failing test you mentioned.

We can do the clean ups that slava mentioned as follow up PRs.
